### PR TITLE
Overlays: allow overlays to be specified as a file

### PR DESCRIPTION
**Pull request by michaelpj**
###### Motivation for this change
This allows overlays to be specified as a file containing a list of overlays in two ways:
- `<nixpkgs-overlays>` can point to a file.
- `~/.confix/nixpkgs/overlays.nix` can be used.

This enables two use-cases which have bothered me (and several other people on IRC):
- It allows you to specify overlays for Nixpkgs with an explicit order. Currently you can do this on NixOS with the `nixpkgs.overlays` option, but Nixpkgs has to use the directory method, which cedes control over what order the overlays are applied in.
- It allows you to conveniently share an overlays specification between a NixOS system configuration and user nixpkgs, by both importing a file as `nixpkgs.overlays` and symlinking it to `~/.config/nixpkgs/overlays.nix`. Currently NixOS uses an explicit list and Nixpkgs uses a directory, which is annoying. (You can use `NIX_PATH`, but I prefer doing it more explicitly.)
    - This is the same approach that you can use for `config.nix`, which makes things nice and consistent.

Based on https://github.com/NixOS/nixpkgs/pull/28082 for documentation.

@nbp 

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/impure.nix [hunk 1]
```diff
@@ -40,18 +40,25 @@ in
...
-        map (n: import (dir + ("/" + n)))
-          (builtins.filter (n: builtins.match ".*\.nix" n != null || pathExists (dir + ("/" + n + "/default.nix")))
-            (attrNames content));
+      pathOverlays = try (if pathExists <nixpkgs-overlays> then <nixpkgs-overlays> else "") "";
+      homeOverlaysDir = homeDir + "/.config/nixpkgs/overlays";
+      homeOverlaysFile = homeDir + "/.config/nixpkgs/overlays.nix";
+      overlays = path:
+        # check if the path is a directory or a file
+        if pathExists (path + "/.")
+        # it's a directory, so the set of overlays from the directory, ordered lexicographically
```
Comment (nbp) [inline pkgs/top-level/impure.nix]: nit: move this comment under the then branch.

File: pkgs/top-level/impure.nix [hunk 2]
```diff
@@ -40,18 +40,25 @@ in
...
+              (attrNames content))
+        # it's a file, so the result is the contents of the file itself
+        else import path;
     in
-      if dirPath != "" then
-        overlays dirPath
-      else if dirCheck dirHome then overlays dirHome
+      if pathOverlays != "" then
+        overlays pathOverlays
+      else if pathExists homeOverlaysFile then overlays homeOverlaysFile
```
Comment (nbp) [inline pkgs/top-level/impure.nix]: What if there is a directory named `overlay.nix`?

File: pkgs/top-level/impure.nix [hunk 3]
```diff
@@ -40,18 +40,25 @@ in
...
+      else if pathExists homeOverlaysDir then overlays homeOverlaysFile
```
Comment (nbp) [inline pkgs/top-level/impure.nix]: What if there is a file named `overlay` without extensions?

**Review by nbp — CHANGES_REQUESTED** [review]
I understand the use case. Now I think I might one alternative and/or an other proposal to offer.

Comments:
 - I like the idea that <nixpkgs-overlays> can be either a nix file, or a directory.
 - I do not like the idea of ~/.config/nixpkgs/overlay.nix , as this defeat the purpose of having overlays.

Alternative:
 - NixOS gives you the ability to set the environment variable, thus you could technically define from NixOS the NIX_PATH environment variable with the extra overlays.

Other proposal:
 - Overlay directories: list of overlays, or overlay directories.
 - Overlay: function with `self: super:`

`overlays` argument, `<nixpkgs-overlay>` are overlay directories. `~/.config/nixpkgs/overlays` is a (fs-)directory of overlay or overlay directories.

Thus you could have a single file in the ~/.config/nixpkgs/overlays/nixos-overlays.nix , which contains the list imported in NixOS.

What do you think?

**Comment by michaelpj** [comment]
Thanks for the comments!

> I do not like the idea of ~/.config/nixpkgs/overlay.nix , as this defeat the purpose of having overlays.

Could you expand on this comment a bit? I don't quite understand how it defeats the point of having overlays.

To guess a bit: currently the selection of which overlays to use is implicit in the filesystem, rather than explicit in the config (except in NixOS). This means that you can alter which overlays you use without having to change your configuration.

The alternative is to configure the list of overlays explicitly (like in NixOS). Personally, I prefer this approach, but I can see the attraction of the implicit version. It's also more consistent with `config.nix` - the configuration is just a file.

However, this PR merely adds the *option* of providing the overlays explicitly, you can still use a directory if you want. The use cases that I mentioned seem to mostly bother people who want to be explicit about everything (like me!).

> NixOS gives you the ability to set the environment variable, thus you could technically define from NixOS the NIX_PATH environment variable with the extra overlays.

Firstly, without the ability to set `<nixpkgs-overlays>` to a file this solves only the second use-case, not the first.

Secondly, I think there are actually three logically distinct ways of setting things. Let's compare with `config.nix`:
- For the NixOS system Nixpkgs: `nixpkgs.config`
- For the user Nixpkgs: `~/.config/nixpkgs/config.nix`
- For the user NIxpkgs of all users of a NixOS system, by the globally set `NIXPKGS_CONFIG` variable (set to `/etc/nix/nixpkgs-config.nix`).
    - Note that I think this should be like `<nixpkgs-overlays>` and come from `NIX_PATH` instead, see https://github.com/NixOS/nixpkgs/issues/28085.

This gives you a couple of options to solve the second use case for `config.nix`:
- Use one file for `config.nix` and then either:
    - Symlink it to `~/.config/nixpkgs/config.nix`; or
    - Symlink it to `/etc/nix/nixpkgs-config.nix`.

Now, I don't think that either of these is obviously superior to the other. The former will work also on non-NixOS systems, but the latter will work for all users. So I think there's a case for supporting both.

Setting `NIX_PATH` on NixOS to include `<nixpkgs-overlays>` would let us use the second approach, which I'm in favour of, but I think it's worth having both.

> Other proposal ....

I think this would work, but I'm not a huge fan, since (among other things) it makes the types of overlays files more complex. We would also need to carefully specify how the "flattening" takes place.

I think I might understand better if I understood why you don't like the file-based approach :)

File: pkgs/top-level/impure.nix (hunk 2 above)
Comment (michaelpj) [inline pkgs/top-level/impure.nix]: Thanks, good point. It should be agnostic for the value of `<nixpkgs-overlays>`, but not for the locations in the home directory.

**Comment by nbp** [comment]
I do not like `~/.config/nixpkgs/overlay.nix` because it implies that a file should be modified, and I am less convinced that we would have tools to handle the content of a file compared to having tools handle the content of a directory.

For example, the rust-overlay can be install with the following script:
https://github.com/mozilla/nixpkgs-mozilla/blob/master/rust-overlay-install.sh

In the Nixpkgs argument, there is a different way of calling Nixpkgs which makes it clear that this is not the ordinary Nixpkgs.

With the NIX_PATH, or -I, there is a deliberate configuration setup by the user.

With the file and the directory, external tools can been adding both of these, without the user noticing the fact that it might be using both.  I think we should probably support both at the same time, where for example the content of the directory is appended to the list provided by `overlay.nix` or the opposite.

**Comment by michaelpj** [comment]
Don't these arguments apply equally to `config.nix`? That's not easy for tools to manage either.

As I said, I can see the value in having the imperative style around - it works nicely for users who are using nixpkgs primarily through `nix-env`. Arguably having to edit `config.nix` is also a bad workflow for those people.

However, there are also people who *want* to write down everything explicitly, even for their user configuration (look at the excitement over NixUP!). `overlays.nix` won't be used unless it's present, so you have to opt-in, at which point you should know what you're doing. I don't think it's too awkward for the rust-overlay installation instructions to say "Run this script; or, if you're using an explicit `overlays.nix`, add the overlay to it".

In short, perhaps this is a feature that's not going to be used by everyone, but I think there are a lot of people around who will want to use it.

**Comment by nbp** [comment]
> Don't these arguments apply equally to config.nix? That's not easy for tools to manage either.

Yes, and that's not a reason to keep making the same mistakes.

My problem is about having precedence between the imperative (overlay directory) vs. the declarative (overlay.nix file) approaches.  This will add more confusion to any user who only wants to use nix in both ways at the same time, and I think we should aim for the least surprises, i-e merging the overlay directory with the overlay.nix file.

**Comment by michaelpj** [comment]
FWIW, I found it very surprising that `overlays.nix` didn't already exist, because it seemed very natural to me to want to specify my overlays declaratively, like `config.nix`. Otherwise the symmetry between the two is very straightforward, and makes me expect a simple file next to `config.nix`.

While I don't like the idea of actually putting the `overlays.nix` file in the `overlays` directory, I do think that *merging* the two sounds plausible. I think it would be reasonable to concatenate:
- `<nixpkgs-overlays>`
- `overlays.nix`
- The overlays in `overlays`
in that order, since that corresponds to roughly the precedence I'd expect (global -> declarative -> imperative). I notice that you discussed this a bit already https://github.com/NixOS/nixpkgs/pull/21243#issuecomment-276750798.

(Maybe we could merge the attribute sets for the various sources of `config.nix` too.)

If we did that, then I think this PR would look very similar to your suggestion of putting `overlays.nix` inside the `overlays` directory, except with more control over ordering (it makes more sense for the declarative overlays to go before all the imperative ones, rather than just before those that come after 'o'!).

I also note that nix files that evaluate to lists are going to be awkward for tools no matter where they are - in your alternate suggestion, what does the tool do with such a file in `overlays`? It seems just as stuck as if it were in `overlays.nix`.

**Comment by nbp** [comment]
> […] I do think that merging the two sounds plausible […]

```nix
  (if pathExists overlayNix then import overlayNix else []) ++
  (if pathExists overlayDir then overlays overlayDir else [])
```

**Comment by nbp** [comment]
I still think that this is nicer to have `<nixpkgs-overlays>` take precedence.

The problem I see with the introduction of the `overlay.nix` file, is that `<nixpkgs-overlays>` was made to be able to just have another directory and use it as a replacement.  Having both sources, i-e overlay.nix and overlay directory, implies that one cannot replace one directory by just another directory with `<nixpkgs-overlays>`.

Thus, maybe, instead of merging them, we should make them exclusive, in a way that we get an error if both the `overlay.nix` file, and the `overlay` directory are present at the same time.

**Comment by michaelpj** [comment]
Okay, leaving the merging to one side because it's tricky and we can do it later.

What do you think about this point that I made?
>I also note that nix files that evaluate to lists are going to be awkward for tools no matter where they are - in your alternate suggestion, what does the tool do with such a file in overlays? It seems just as stuck as if it were in overlays.nix.

If that's true then it looks like your alternative has the same problem with tools, in which case I think my suggestion is the simplest thing that gets us some of what we want.

**Comment by michaelpj** [comment]
> You consider the aggregate as a single unit, which can be installed/removed. For example nixpkgs-mozilla repository would benefit from becoming such aggregate in the future, instead of doing complex work-around to redo the extend logic for such aggregation of overlays.

Right, but consider our example use case. As I understand it, you suggested that you create a file `overlays/nixos-overlays.nix` containing the overlays used in NixOS. But then these overlays can only be managed as a unit, even if they are logically separate (which they might well be, there is no particular reason why the set of overlays used in your NixOS configuration should be monolithic). But then this can't be managed. So a user who wants to have the same set of overlays in NixOS and outside isn't going to be able to use the tool to manage them, regardless.

Although, I wonder if it would be so hard to manage `overlays.nix` files if they were in a relatively simple format. I'm thinking of how Emacs will manage some parts of your `init.el` for you, despite that being code.

I actually find the argument about overlays being annoying to compose quite convincing. If it were easy to compose them then the mozilla overlay could just do that, but since we're pre-nixpkgs we can't provide an nice library function. So I guess it's somewhat nice to be able to distribute a list of overlays as well.

However, that's an argument for allowing files that are lists of overlays in `overlays`, I don't think it's necessarily an argument for not having `overlays.nix`. `overlays.nix` is still the only thing that gives you complete control over what overlays are used, and what order they're applied in. Anything that uses the directory sacrifices that.

**Comment by nbp** [comment]
> I actually find the argument about overlays being annoying to compose quite convincing. If it were easy to compose them then the mozilla overlay could just do that, but since we're pre-nixpkgs we can't provide an nice library function. So I guess it's somewhat nice to be able to distribute a list of overlays as well.

This is not actually true, because we can already use `super`, but this far from being easy to understand.

> However, that's an argument for allowing files that are lists of overlays in overlays, I don't think it's necessarily an argument for not having overlays.nix. overlays.nix is still the only thing that gives you complete control over what overlays are used, and what order they're applied in. Anything that uses the directory sacrifices that.

As I mention, my issue is that they are both controlled with the file system, and every program can mutate it without approval.  Having one way being added which takes precedence over the other will lead to weird errors, which after installing an overlay, a user will not see the previously installed overlays.

This, alone, leads to either merging them or have exclusively only one of them.

Then, the idea of <nixpkgs-overlays>, was to be able to replace one directory by another. Merging overlay.nix with the overlay directory implies that the promise of <nixpkgs-overlays> no longer hold.

Thus, we should accept one (x)or the other, but not both.

**Comment by michaelpj** [comment]
> As I mention, my issue is that they are both controlled with the file system, and every program can mutate it without approval. Having one way being added which takes precedence over the other will lead to weird errors, which after installing an overlay, a user will not see the previously installed overlays.

This isn't true with my original proposal, I think? Either you're using `overlays.nix`, in which case imperative overlays management is effectively "off" (and the tool can know this), or you're not, and installation is never going to create it.

> Then, the idea of , was to be able to replace one directory by another. Merging overlay.nix with the overlay directory implies that the promise of no longer hold.

Yeah, I'm not sure about the merging, which is why I'd like to focus on the original proposal. Note that `<nixpkgs-overalys>` can still be a directory, so you can still replace a directory with a directory (or file) in that way.

**Comment by nbp** [comment]
> Either you're using overlays.nix, in which case imperative overlays management is effectively "off" (and the tool can know this), or you're not, and installation is never going to create it.

And my point is, that we should raise an error if both exist, such that if a bad install script create the other  file, the user can blame the script for introducing such error instead of blaming Nixpkgs for not reading the installed files.

**Comment by michaelpj** [comment]
Okay, I admit that I can't think of a good reason to have both `overlays.nix` and an `overlays` directory.

So the proposed logic would look like this:
- First, `<nixpkgs-overlays>` if it exists, is used. The path can be either a file or a directory.
- Then we look to see if `overlays.nix` or `overlays` exists. If both do then it's an error. If exactly one exists, then it is used.

Does that look right?

**Comment by nbp** [comment]
That sounds good to me.
Thanks for this discussion.

**Comment by michaelpj** [comment]
Likewise, very productive :)

I'll implement that and get back to you.

**Comment by michaelpj** [comment]
Updated as discussed.

File: pkgs/top-level/impure.nix [hunk 4]
```diff
@@ -40,18 +40,34 @@ in
...
-          (builtins.filter (n: builtins.match ".*\.nix" n != null || pathExists (dir + ("/" + n + "/default.nix")))
-            (attrNames content));
+      isDir = path: pathExists (path + "/.");
+      pathOverlays = try <nixpkgs-overlays> "";
+      homeOverlaysFile = homeDir + "/.config/nixpkgs/overlays.nix";
+      homeOverlaysDir = homeDir + "/.config/nixpkgs/overlays";
+      overlays = path:
+        # check if the path is a directory or a file
+        if isDir path
+        then
```
Comment (nbp) [inline pkgs/top-level/impure.nix]: nit: move "then" to the previous line.

File: pkgs/top-level/impure.nix [hunk 5]
```diff
@@ -40,18 +40,34 @@ in
...
+        else 
+          # it's a file, so the result is the contents of the file itself
+          import path;
     in
-      if dirPath != "" then
-        overlays dirPath
-      else if dirCheck dirHome then overlays dirHome
+      if pathOverlays != "" && pathExists pathOverlays then overlays pathOverlays
+      else if pathExists homeOverlaysFile && pathExists homeOverlaysDir then 
+        throw ("Precisely one of " + homeOverlaysFile + " and " + homeOverlaysDir + " must exist")
```
Comment (nbp) [inline pkgs/top-level/impure.nix]: nit: This message is not clear, and it should at least suggest one solution to solve the problem.

```nix
  throw ''
    Nixpkgs cannot handle overlays when both ${homeOverlaysFile} and ${homeOverlaysDir} exist.
    Try to remove one of them and try again.
  ''
```

File: doc/overlays.xml [hunk 6]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
 <section xml:id="sec-overlays-install">
 <title>Installing overlays</title>
 
-<para>The list of overlays is determined as follows:
+<para>The list of overlays is determined as follows.</para>
 
-<orderedlist>
+<para>Firstly, if an <varname>overlays</varname> argument to the nixpkgs function itself is given,
+then that is used. This can be passed explicitly when importing nipxkgs, for example 
+<literal>import &lt;nixpkgs> { overlays = [ overlay1 overlay2 ] }</literal>.</para>
```
Comment (nbp) [inline doc/overlays.xml]: nit: expected semi-colon after token ']', found '}'.

File: doc/overlays.xml [hunk 7]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
 
-  <listitem>
-    <para>First, if an <varname>overlays</varname> argument to the nixpkgs function itself is given,
-    then that is used. This can be passed explicitly when importing nipxkgs, for example 
-    <literal>import &lt;nixpkgs> { overlays = [ overlay1 overlay2 ] }</literal>.</para>
+<para>If the overlays argument is not provided explicitly, we look for overlays in a path. The path
```
Comment (nbp) [inline doc/overlays.xml]: nit: `the <varname>overlays</varname> argument`

File: doc/overlays.xml [hunk 8]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
-    <para>On a NixOS system the value of the <literal>nixpkgs.overlays</literal> option, if present, 
-    is passed to the system Nixpkgs in this way. Note that this does not affect the overlays for
-    non-NixOS operations (e.g. <literal>nix-env</literal>), which are looked up independently.</para>
-  </listitem>
+<orderedlist>
 
   <listitem>
-    <para>Otherwise, if the Nix path entry <literal>&lt;nixpkgs-overlays></literal> exists and is a 
-    directory, then the result is the set of overlays found in that directory, ordered lexicographically.</para> 
+    <para>First, if the Nix path entry <literal>&lt;nixpkgs-overlays></literal> exists, then it is used.</para>
```
Comment (nbp) [inline doc/overlays.xml]: nit: `Firstly` then `First`.  and having an ordered list with only 2 items.  I think it makes more sense to have all 3 cases under the same ordered list.

File: doc/overlays.xml [hunk 9]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
 
 </orderedlist>
 </para>
 
-<para>For the second and third options, overlays are extracted from the given directory as files, 
-directories containing a <filename>default.nix</filename>, or symlinks to one of those.</para>
+<para>If the path is a file, then the file is imported as a Nix expression and used as the list of
+overlays.</para>
+
+<para>If the path is a directory, then we take the set of children of the directory and order it
```
Comment (nbp) [inline doc/overlays.xml]: nit: s/set of children/content/

File: doc/overlays.xml [hunk 10]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
+lexicographically. Then we attempt to extract an overlay from each by:
```
Comment (nbp) [inline doc/overlays.xml]: nit: … content of the directory, order it lexicographically, and interpret them as overlays by:

File: doc/overlays.xml [hunk 11]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
+<orderedlist>
```
Comment (nbp) [inline doc/overlays.xml]: nit: unordered list?

File: doc/overlays.xml [hunk 12]
```diff
@@ -17,42 +17,55 @@ if multiple layers override the same package.</para>
...
+  <listitem>
+    <para>Importing the file, if it is a <literal>.nix</literal> file.</para>
+  </listitem>
+  <listitem>
+    <para>Importing a newsted <filename>default.nix</filename> file, if it is a directory.</para>
```
Comment (nbp) [inline doc/overlays.xml]: nit: s/newsted/nested/, I would prefer "top-level", as we are not recursively looking for a nested file named default.nix.

**Review by nbp — APPROVED** [review]
Apply the following nits, and I think this would be good to be merged.

Nice work!

File: doc/overlays.xml (hunk 10 above)
Comment (michaelpj) [inline doc/overlays.xml]: I'm going to keep the "attempt", since we ignore cases that can't be interpreted in this way.

**Comment by michaelpj** [comment]
Thanks for your comments, I've applied them.

Do you think this should be announced on the mailing list or mentioned in the release notes?

**Comment by nbp** [comment]
Looking at the number of likes , feel free to announce.
I think adding it to the release notes of 17.09 would be good.

**Comment by michaelpj** [comment]
I've added a release note.

**Comment by FRidh** [comment]
> - Overlay: function with `self: super:`

I would like to see this in. Typically, my expressions are pure; that is, I fetch a version of Nixpkgs with `fetchTarball` and a hash, and when I want to modify Nixpkgs I want to provide an overlay, but not necessarily as a different file. Therefore, I think it should be possible to just pass in a `self: super:` directly as well.

File: pkgs/top-level/impure.nix [hunk 13]
```diff
@@ -40,18 +40,36 @@ in
...
   overlays ? let
-      dirPath = try (if pathExists <nixpkgs-overlays> then <nixpkgs-overlays> else "") "";
-      dirHome = homeDir + "/.config/nixpkgs/overlays";
-      dirCheck = dir: dir != "" && pathExists (dir + "/.");
-      overlays = dir:
-        let content = readDir dir; in
-        map (n: import (dir + ("/" + n)))
-          (builtins.filter (n: builtins.match ".*\.nix" n != null || pathExists (dir + ("/" + n + "/default.nix")))
-            (attrNames content));
+      isDir = path: pathExists (path + "/.");
```
Comment (solson) [inline pkgs/top-level/impure.nix]: Is it just me, or does this not work at all if `path` is actually a path (according to `builtins.typeOf`)? For a string like `"/file.nix"`, it does indeed check `"/file.nix/."` and return `false`, but for a path like `/file.nix`, the `path + "/."` expression just returns `/file.nix` again, and `isDir` returns `true`!

This currently means I can't actually use `<nixpkgs-overlays>` pointing at a file at all, because `isDir` returns `true`, it tries to read the file as a directory, and it crashes.

I found another place in nixpkgs that does its own `isDir` in a seemingly more robust way [here](https://github.com/NixOS/nixpkgs/blob/b760158c283755e5d68a253eca0010262d0b6aa4/lib/filesystem.nix#L39-L43). It may be best to make it a top-level lib function rather than reimplementing it in multiple places.

File: pkgs/top-level/impure.nix [hunk 14]
```diff
@@ -40,18 +40,36 @@ in
...
+          # it's a file, so the result is the contents of the file itself
+          import path;
     in
-      if dirPath != "" then
-        overlays dirPath
-      else if dirCheck dirHome then overlays dirHome
+      if pathOverlays != "" && pathExists pathOverlays then overlays pathOverlays
+      else if pathExists homeOverlaysFile && pathExists homeOverlaysDir then 
+        throw ''
+          Nixpkgs overlays can be specified with ${homeOverlaysFile} or ${homeOverlaysDir}, but not both.
```
Comment (tex) [inline pkgs/top-level/impure.nix:62]: Yeah, this is not very helpful message. I am trying to get rust nightly and got this message. Now finding with google lead me here. I still do not know what I should do.

File: pkgs/top-level/impure.nix (hunk 13 above)
Comment (TyberiusPrime) [reply pkgs/top-level/impure.nix:43]: 6 years later, I do second solsons opinion on this, that only works on strings, not actual paths.
